iotdb/src/main/java/org/apache/iotdb/db/engine/cache/TsFileMetaDataCache.java

```diff
@@ -31,67 +32,113 @@
 public class TsFileMetaDataCache {
 
   private static final Logger logger = LoggerFactory.getLogger(TsFileMetaDataCache.class);
+  private static final IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+
+  private static final long MEMORY_THRESHOLD_IN_B = (long) (0.25 * config
+      .getAllocateMemoryForRead());
   /**
-   * key: The file seriesPath of tsfile.
+   * key: The file seriesPath of tsfile. value: TsFileMetaData
    */
-  private ConcurrentHashMap<String, TsFileMetaData> cache;
+  private LruLinkedHashMap<String, TsFileMetaData> cache;
   private AtomicLong cacheHintNum = new AtomicLong();
   private AtomicLong cacheRequestNum = new AtomicLong();
 
+  /**
+   * estimate size of a deviceIndexMap entry in TsFileMetaData.
+   */
+  private long deviceIndexMapEntrySize = 0;
+  /**
+   * estimate size of measurementSchema entry in TsFileMetaData.
+   */
+  private long measurementSchemaEntrySize = 0;
+  /**
+   * estimate size of version and CreateBy in TsFileMetaData.
+   */
+  private long versionAndCreatebySize = 10;
+
   private TsFileMetaDataCache() {
-    cache = new ConcurrentHashMap<>();
+    cache = new LruLinkedHashMap<String, TsFileMetaData>(MEMORY_THRESHOLD_IN_B, true) {
+      @Override
+      protected long calEntrySize(String key, TsFileMetaData value) {
+        if (deviceIndexMapEntrySize == 0 && value.getDeviceMap().size() > 0) {
+          deviceIndexMapEntrySize = RamUsageEstimator
+              .sizeOf(value.getDeviceMap().entrySet().iterator().next());
+        }
+        if (measurementSchemaEntrySize == 0 && value.getMeasurementSchema().size() > 0) {
+          measurementSchemaEntrySize = RamUsageEstimator
+              .sizeOf(value.getMeasurementSchema().entrySet().iterator().next());
+        }
+        long valueSize = value.getDeviceMap().size() * deviceIndexMapEntrySize
+            + measurementSchemaEntrySize * value.getMeasurementSchema().size()
+            + versionAndCreatebySize;
+        return key.length() + valueSize;
```

Comment:
String.length returns the number of characters and a Java char consumes 2 bytes.

Reply:
fixed